File: 173_binary_search_tree_iterator.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class BSTIterator {
    Stack<TreeNode> stack;

    public BSTIterator(TreeNode root) {
        stack = new Stack<>();
        TreeNode curr = root;
        while (curr != null) {
            stack.push(curr);
            curr = curr.left;
        }
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return stack.size() != 0;
    }

    /** @return the next smallest number */
    public int next() {
        TreeNode node = stack.pop();

        TreeNode curr = node.right;
        while (curr != null) {
            stack.push(curr);
            curr = curr.left;
        }

        return node.val;
    }

//     private Queue<Integer> queue;

//     public BSTIterator(TreeNode root) {
//         queue = new LinkedList<>();
//         iterateTree(queue, root);
//     }

//     private void iterateTree(Queue<Integer> queue, TreeNode node) {
//         if (node == null) return;

//         iterateTree(queue, node.left);
//         if (node != null) queue.offer(node.val);
//         iterateTree(queue, node.right);
//     }

//     /** @return whether we have a next smallest number */
//     public boolean hasNext() {
//         return ! queue.isEmpty();
//     }

//     /** @return the next smallest number */
//     public int next() {
//         if (hasNext()) return queue.poll();
//         else return 0;
//     }
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = new BSTIterator(root);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: python/173_binary_search_tree_iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:

    def __init__(self, root: TreeNode):
        self.stack = []
        if root:
            while root:
                self.stack.append(root)
                root = root.left

    def next(self) -> int:
        """
        @return the next smallest number
        """
        top = self.stack.pop()
        right = top.right
        while right:
            self.stack.append(right)
            right = right.left

        return top.val

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return self.stack


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
